login/profile.go

```diff
@@ -0,0 +1,129 @@
+type KeycloakUserProfile struct {
+	ID         *string                        `json:"id,omitempty"`
+	CreatedAt  int64                          `json:"createdTimestamp,omitempty"`
+	Username   *string                        `json:"id,omitempty"`
+	FirstName  *string                        `json:"firstName,omitempty"`
+	LastName   *string                        `json:"lastName,omitempty"`
+	Email      *string                        `json:"email,omitempty"`
+	Attributes *KeycloakUserProfileAttributes `json:"attributes,omitempty"`
+}
+
+// KeycloakUserProfileAttributes represents standard Keycloak profile payload Attributes
+type KeycloakUserProfileAttributes struct {
+	Bio      *string `json:"bio,omitempty"`
+	URL      *string `json:"url,omitempty"`
+	ImageURL *string `json:"image_url,omitempty"`
+}
+
+//KeycloakUserProfileResponse represents the user profile api response from keycloak
+type KeycloakUserProfileResponse struct {
+	ID                         *string                      `json:"id"`
+	CreatedTimestamp           *int64                       `json:"createdTimestamp"`
+	Username                   *string                      `json:"username"`
+	Enabled                    *bool                        `json:"enabled"`
+	Totp                       *bool                        `json:"totp"`
+	EmailVerified              *bool                        `json:"emailVerified"`
+	FirstName                  *string                      `json:"firstName"`
+	LastName                   *string                      `json:"lastName"`
+	Email                      *string                      `json:"email"`
+	Attributes                 *KeycloakUserProfileResponse `json:"attributes"`
+	DisableableCredentialTypes []*string                    `json:"disableableCredentialTypes"`
+	RequiredActions            []interface{}                `json:"requiredActions"`
+}
+
+type KeycloakUserProfileResponseAttributes struct {
+	Bio      []*string `json:"bio"`
+	ImageURL []*string `json:"image_url"`
+	URL      []*string `json:"url"`
+}
+
+// NewKeycloakUserProfile creates a new keycloakUserProfile instance.
+func NewKeycloakUserProfile(firstName, lastName, email, bio, url, imageURL *string) *KeycloakUserProfile {
+	return &KeycloakUserProfile{
+		FirstName: firstName,
+		LastName:  lastName,
+		Email:     email,
+		Attributes: &KeycloakUserProfileAttributes{
+			Bio:      bio,
+			URL:      url,
+			ImageURL: imageURL,
+		},
+	}
+}
+
+// UserProfileService describes what the services need to be capable of doing.
+type UserProfileService interface {
+	Update(keycloakUserProfile *KeycloakUserProfile, accessToken string, keycloakProfileURL string) error
+	Get(accessToken string, keycloakProfileURL string) (*KeycloakUserProfileResponse, error)
+}
+
+// KeycloakUserProfileClient describes the interface between platform and Keycloak User profile service.
+type KeycloakUserProfileClient struct {
+	client *http.Client
```

Comment:
Sure. Introducing KeycloakUserProfileClient as an implementation of UserProfileService makes perfect sense. I just not sure why to keep a pointer to the http.Client all the time (this service lives forever) instead of creating a new instance when you need it (when updating Profile) and let garbage collector to clean up.